### PR TITLE
[NEW] Implement sequential (grouped) messages

### LINKED ISSUE #617: [NEW] Implement sequential (grouped) messages
Some messages should be grouped together when the same user posts in sequence.

The web app makes some exceptions like images which do not merge.
I haven't investigated which other ones don't merge yet.

### PR DESCRIPTION
**Pull request by cardoso**
@RocketChat/ios

Closes #617

### CODE REVIEW

**Comment by cardoso** [comment]
| Preview (normal) | Preview (debug) |
|-|-|
|||

File: Rocket.Chat/Views/Cells/Chat/ChatMessageCell.swift [hunk 1]
```diff
@@ -55,13 +55,13 @@ final class ChatMessageCell: UICollectionViewCell {
...
 
-    static func cellMediaHeightFor(message: Message, grouped: Bool = true) -> CGFloat {
+    static func cellMediaHeightFor(message: Message, sequential: Bool = true) -> CGFloat {
         let fullWidth = UIScreen.main.bounds.size.width
         var total = UILabel.heightForView(
             MessageTextCacheManager.shared.message(for: message)?.string ?? "",
             font: UIFont.systemFont(ofSize: 15),
             width: fullWidth - 62
-        ) + 35
+            ) + (sequential ? 7 : 35)
```
Comment (rafaelks) [inline Rocket.Chat/Views/Cells/Chat/ChatMessageCell.swift]: Can you reduce this extra espace here?

File: Rocket.Chat/Views/Cells/Chat/ChatMessageCell.swift [hunk 2]
```diff
@@ -87,6 +87,19 @@ final class ChatMessageCell: UICollectionViewCell {
...
+    // MARK: Sequential
+    @IBOutlet weak var labelUsernameHeightConstraint: NSLayoutConstraint!
+    @IBOutlet weak var labelDateHeightConstraint: NSLayoutConstraint!
+    @IBOutlet weak var avatarContainerHeightConstraint: NSLayoutConstraint!
+
+    var sequential: Bool = false {
+        didSet {
+            avatarContainerHeightConstraint.constant = sequential ? 0 : 35
+            labelUsernameHeightConstraint.constant = sequential ? 0 : 21
+            labelDateHeightConstraint.constant = sequential ? 0 : 21
```
Comment (rafaelks) [inline Rocket.Chat/Views/Cells/Chat/ChatMessageCell.swift]: What about even not adding values to avatar/timestamp/username labels? This way we can avoid some processing into the app.

File: Rocket.Chat/Models/Message.swift [hunk 3]
```diff
@@ -31,6 +31,14 @@ enum MessageType: String {
     case subscriptionRoleRemoved = "subscription-role-removed"
     case roomArchived = "room-archived"
     case roomUnarchived = "room-unarchived"
+
+    var sequential: Bool {
+        if self == .image {
+            return false
+        }
```
Comment (rafaelks) [inline Rocket.Chat/Models/Message.swift]: I believe the type `video` could also be used to return `false`. What do you think?

Comment (cardoso) [reply Rocket.Chat/Models/Message.swift]: I can add that, but videos don't even show up here. Are they implemented?

File: Rocket.Chat/Views/Cells/Chat/ChatMessageCell.swift (hunk 2 above)
Comment (cardoso) [reply Rocket.Chat/Views/Cells/Chat/ChatMessageCell.swift]: Yeah, i thought about that. I'll give it a second try, but as far as i know views with height/width == 0 are ignored by the rendering pipeline.

File: Rocket.Chat/Models/Message.swift (hunk 3 above)
Comment (rafaelks) [reply Rocket.Chat/Models/Message.swift]: They show up, you need to upload a video file.

File: Rocket.Chat/Views/Cells/Chat/ChatMessageCell.swift (hunk 2 above)
Comment (rafaelks) [reply Rocket.Chat/Views/Cells/Chat/ChatMessageCell.swift]: Yes, they are. But we're still constructing the avatar here: https://github.com/RocketChat/Rocket.Chat.iOS/pull/618/files#diff-535132aa8b35719fb1bcbd5602ca482eR193.

Comment (cardoso) [reply Rocket.Chat/Views/Cells/Chat/ChatMessageCell.swift]: #[Prevent unnecessary computations for sequential cells](https://github.com/RocketChat/Rocket.Chat.iOS/pull/618/commits/987713cafcea9be45ee0558466ac337cb1a696f8)

File: Rocket.Chat/Models/Message.swift (hunk 3 above)
Comment (cardoso) [reply Rocket.Chat/Models/Message.swift]: #[Prevent video being sequential message](https://github.com/RocketChat/Rocket.Chat.iOS/pull/618/commits/9618ce82d685d4bfa45af35073d6d7613228654c)

**Comment by rafaelks** [comment]
Thank you for the changes, but looks like the usage of this changes still has lots of bugs, look at this GIF...

**Comment by cardoso** [comment]
Thanks for spotting that. I'll see if i can fix it, but i think demo.rocket.chat is down :/

**Comment by rafaelks** [comment]
@cardoso Yes, it is. You can use `http://rocket.chat/deploy` to deploy a new server to test, it's free now.

**Comment by cardoso** [comment]
Rookie mistake, forgot to reset the property in prepareForReuse

#[Reset sequential cell property in prepareForReuse (bugfix)](https://github.com/RocketChat/Rocket.Chat.iOS/pull/618/commits/b339f846c361c131eb20c21d2be529627c8852d7)

**Comment by rafaelks** [comment]
@cardoso Please, let me know when this PR is ready to review again, OK? Thanks!

**Comment by cardoso** [comment]
I think it's ready. I couldn't be 100% sure which types of messages are sequential or not, but i t can be changed easily later.

**Comment by rafaelks** [comment]
@cardoso Looking very good from my tests. Tested lots of different situations, with different kind of messages and different devices.

I've one suggestion, see what you think: let's also include the user information if the difference of time between one message and another is more than 30 minutes? Let me explain you why I think that... if I'm reading a channel and one user sent 5 messages, I read everything and I'm done. I come back and there's one new message from the same user, 1 hour after the others. I think adding user avatar/username in this situation will be more effective for me to read the new message. What do you think?

**Comment by cardoso** [comment]
Definitely a fast tweak, I'll look into it

**Pull request merged**